pennylane/resources/second_quantization.py

```diff
@@ -0,0 +1,216 @@
+# Copyright 2018-2022 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This module contains the functions needed for resource estimation with the double factorization
+method.
+"""
+# pylint: disable= too-many-arguments
+from pennylane import numpy as np
+
+
+def estimation_cost(norm, error):
+    r"""Return the number of calls to the unitary needed to achieve the desired error in phase
+    estimation.
+
+    Args:
+        norm (float): 1-norm of a second-quantized Hamiltonian
+        error (float): target error in the algorithm
+
+    Returns:
+        int: number of calls to unitary
+
+    **Example**
+
+    >>> cost = estimation_cost(72.49779513025341, 0.001)
+    >>> print(cost)
+    113880
+    """
+    return int(np.ceil(np.pi * norm / (2 * error)))
+
+
+def qrom_cost(constants):
+    r"""Return the number of Toffoli gates and the expansion factor needed to implement a QROM.
+
+    The complexity of a QROM computation in the most general form is given by
+    [`arXiv:2011.03494 <https://arxiv.org/abs/2011.03494>`_]
+
+    .. math::
+
+        \text{cost} = \left \lceil \frac{a + b}{k} \right \rceil + \left \lceil \frac{c}{k} \right
+        \rceil + d \left ( k + e \right ),
+
+    where :math:`a, b, c, d, e` are constants that depend on the nature of the QROM implementation
+    and the expansion factor :math:`k` is an integer power of two, :math:`k = 2^n`, that minimizes
+    the cost. This function computes the optimum :math:`k` and the minimum cost for a QROM
+    specification.
+
+    To obtain the optimum values of :math:`k`, we first assume that the cost function is continues
+    and use differentiation to obtain the value of :math:`k` that minimizes the cost. This value of
+    :math:`k` is not necessarily an integer power of 2. We then obtain the value of :math:`n` as
+    :math:`n = \log_2(k)` and compute the cost for
+    :math:`n_{int}= \left \{\left \lceil n \right \rceil, \left \lfloor n \right \rfloor \right \}`.
+    The value of :math:`n_{int}` that gives the smaller cost is used to compute the optimim
+    :math:`k`.
+
+    Args:
+        constants (tuple[float]): constants determining a QROM
```

Comment:
How are these constants determined? Will the user know how to obtain them?

Reply:
This function is not user-facing and is private now.